### ISSUE TITLE
No improvements after several epochs

### ISSUE THREAD

**Issue by abbyDC** [jakeret/tf_unet#122]
Hello! I've been working on a project on stroke lesion segmentation. I've tried it with 14 inputs and 14 ground truths but it only shows an illuminated version of the input as its prediction. Here is my code. (I also used dice coefficient for cost function)

```
# coding: utf-8

from __future__ import division, print_function
import numpy as np

# In[2]:
from tf_unet import image_gen
from tf_unet import unet
from tf_unet import util
from tf_unet import image_util

# In[3]:
# nx = 572
# ny = 572


# In[4]:
generator = image_util.ImageDataProvider('isles/*', data_suffix=".png", mask_suffix='_mask.png')
generator.n_class = 3

# In[5]:
#y_test, x_test = generator(1)

# In[6]:
# ax[0].imshow(x_test[0,...,0], aspect="auto")
# ax[1].imshow(y_test[0,...,1], aspect="auto")

# In[7]:
net = unet.Unet(channels=generator.channels, n_class=generator.n_class, layers=3, features_root=64)

# In[8]:
trainer = unet.Trainer(net, optimizer="adam")

# In[9]:
path = trainer.train(generator, "./unet_trained", training_iters=32, epochs=30, display_step=2)

# In[22]:
x_test, y_test = generator(1)

prediction = net.predict("./unet_trained/model.cpkt", x_test)

unet.error_rate(prediction, util.crop_to_shape(y_test, prediction.shape))
img = util.combine_img_prediction(x_test,y_test,prediction)
util.save_image("isles_problem.png")
```

![screenshot from 2017-10-23 08_25_51](https://user-images.githubusercontent.com/15419597/31868465-d44b2552-b7d1-11e7-818a-4ef44ab041ec.png)

**Comment by jakeret** [comment]
This is an indication that the network is not learning anything. Is the loss decreasing? 
Have you checked that the data from the `generator` is correct? Maybe you need to add some data preprocessing (normalization, augmentation)

**Comment by abbyDC** [comment]
I used dice coefficient for the cost function and I observed that the loss doesn't change that much. It's in the atound the value of -0.1252. I'm kinda new here so may I know how to check if the data from the generator is correct? Thank you!

**Comment by jakeret** [comment]
Try to plot the data you get with 
`x_test, y_test = generator(1)` 
and check the range of the values.

**Comment by abbyDC** [comment]
hello! I tried to plot it using matplotlib but it gives me an error "Value error: x and y can be no greater than 2-D, but have shapes (1,) and (1, 316, 298, 3)". May I ask what does the generator(1) supposed to do? I just patterned it with the demo problems.

**Comment by jakeret** [comment]
`generator(1)` means that one image should be returned (`generator(100)` would return `(100, 316, 298, 3)` .

Anyway, if your data should be grayscaled but you get an RGB image, that means something is not quite right with the loading. `tf_unet` is not doing any kind of automatic transformation vom color to rgb. This is something you have to implement

**Comment by abbyDC** [comment]
Thanks for all the help so far @jakeret truly appreciate it! :) I was able to preprocess my images to make sure they are grayscaled before unet reads them. I've tried running the code but there were no significant differences with the first outputs I got. The predicted images were still illuminated and I wasn't getting the output I wanted which was a segmented version of the input.

**Comment by JS-phine** [comment]
Hi @abbyDC,
I'm encountering the same issue that you are describing here. :(
How did you manage to solve it? Would you mind sharing your code with me?
